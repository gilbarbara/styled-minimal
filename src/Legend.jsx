import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

import { getTheme, isDefined, px } from './utils/helpers';
import { baseStyles } from './utils/system';

import Box from './Box';

const styles = (props: Object): string => {
  const { color: cl, fontFamily, fontWeight: fw, lineHeight, mb } = props;
  const { color, fontWeight, marginBottom } = getTheme(props, 'legend');

  return css`
    color: ${cl || color};
    display: block;
    font-family: ${fontFamily || 'inherit'};
    font-weight: ${fw || fontWeight};
    line-height: ${lineHeight || baseStyles.lineHeight};
    margin-bottom: ${px(isDefined(mb) ? mb : marginBottom)};
    white-space: nowrap;
  `;
};

const Legend = styled(Box)`
  ${styles};
`;

Legend.displayName = 'Legend';

Legend.propTypes = {
  as: PropTypes.string,
  children: PropTypes.node.isRequired,
  ...Box.basePropTypes,
};

Legend.defaultProps = {
  as: 'legend',
};

export default Legend;
