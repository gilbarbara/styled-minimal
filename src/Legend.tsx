import * as React from 'react';
import styled, { css } from 'styled-components';

import { getTheme, px } from './utils/helpers';
import { baseStyles, lineHeightStyles, sx, system } from './utils/system';

import { LegendProps } from './types';

export const StyledLegend = styled('legend')<LegendProps>(
  props => {
    const { color, marginBottom } = getTheme(props, 'legend');

    return css`
      ${baseStyles};
      color: ${color};
      display: block;
      font-family: inherit;
      line-height: ${lineHeightStyles};
      margin-bottom: ${px(marginBottom)};
      white-space: nowrap;
    `;
  },
  sx,
  system,
);

export const Legend = React.forwardRef<HTMLLegendElement, LegendProps>((props, ref) => (
  <StyledLegend ref={ref} {...props} />
));
