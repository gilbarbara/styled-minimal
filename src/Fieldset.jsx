import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

import { px, spacer, themeGet } from './utils/helpers';

import Box from './Box';
import Legend from './Legend';

const styles = (props: Object): string => {
  const { inline } = props;
  const {
    borderColor,
    borderRadius,
    padding,
    marginBottom,
  } = themeGet(props, 'fieldset');

  return css`
    border: 1px solid ${borderColor};
    border-radius: ${px(borderRadius)};
    margin-bottom: ${px(marginBottom)};
    padding: ${px(padding)};
    text-align:left;

    > *:not(legend) {
      ${inline ? 'display: inline-block;' : ''};
      
      + * {
        ${inline ? `margin-left: ${spacer(2)} ;` : ''};
      }
    }
  `;
};

export const Fieldset = styled(Box)`
  ${styles};
  
  ${Legend} {
    margin: 0;
  }
`;

Fieldset.displayName = 'Fieldset';

Fieldset.propTypes = {
  as: PropTypes.string,
  children: PropTypes.node.isRequired,
  inline: PropTypes.bool,
  ...Box.propTypes,
};

Fieldset.defaultProps = {
  as: 'fieldset',
  inline: false,
};

export default Fieldset;
