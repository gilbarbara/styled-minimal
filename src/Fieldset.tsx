import * as React from 'react';
import styled, { css } from 'styled-components';

import { getTheme, px } from './utils/helpers';
import { baseStyles, sx, system } from './utils/system';

import { FieldsetProps } from './types';

import { Legend, StyledLegend } from './Legend';

export const StyledFieldset = styled('fieldset')<FieldsetProps>(
  props => {
    const { borderColor, borderRadius, padding, marginBottom } = getTheme(props, 'fieldset');

    return css`
      ${baseStyles};
      border: 1px solid ${borderColor};
      border-radius: ${px(borderRadius)};
      margin-bottom: ${px(marginBottom)};
      padding: ${px(padding)};
      text-align: left;

      ${StyledLegend} {
        margin: 0;
      }
    `;
  },
  sx,
  system,
);

export const Fieldset = React.forwardRef<HTMLFieldSetElement, FieldsetProps>(
  ({ children, label, ...rest }: FieldsetProps, ref) => {
    return (
      <StyledFieldset ref={ref} {...rest}>
        {label && <Legend>{label}</Legend>}
        {children}
      </StyledFieldset>
    );
  },
);
