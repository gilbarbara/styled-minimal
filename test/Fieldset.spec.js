import React from 'react';
import { mount } from 'enzyme';

import Fieldset from '../src/Fieldset';
import Legend from '../src/Legend';

describe('Fieldset', () => {
  const wrapper = mount(
    <Fieldset>
      <Legend>Title</Legend>
    </Fieldset>
  );

  it('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
