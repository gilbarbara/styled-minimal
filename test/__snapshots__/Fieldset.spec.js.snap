// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fieldset should render properly 1`] = `
.c2 {
  box-sizing: border-box;
  color: #2E2E2E;
  display: block;
  font-family: inherit;
  font-weight: bold;
  line-height: 1.4;
  margin-bottom: 8px;
  white-space: nowrap;
}

.c0 {
  box-sizing: border-box;
  border: 1px solid #ccc;
  border-radius: 4px;
  margin-bottom: 16px;
  padding: 8px;
  text-align: left;
}

.c0 .c1 {
  margin: 0;
}

<Fieldset
  as="fieldset"
  inline={false}
>
  <fieldset
    className="c0"
  >
    <Legend
      as="legend"
    >
      <legend
        className="c1 c2"
      >
        Title
      </legend>
    </Legend>
  </fieldset>
</Fieldset>
`;
